[Tooling] Add localization automation

iangmaia opened the pull request:
Related: [AINFRA-765: Integrate new Fluent conversion tool into wordpress-rs](https://linear.app/a8c/issue/AINFRA-765/integrate-new-fluent-conversion-tool-into-wordpress-rs)

**Note**: The `fluent-tools` Gem used in the PR is currently using the Gem directly from the branch [Add Ruby Gem wrapper / CLI interface and CI setup](https://github.com/Automattic/fluent-rust-tools/pull/3). 🎗️ We need to update it once the Rust tool is fully merged.

This PR supersedes https://github.com/Automattic/wordpress-rs/pull/756.

## What does it do?
This PR integrates the new [Fluent conversion tool](https://github.com/Automattic/fluent-rust-tools) Gem into `wordpress-rs` to enable a complete translation workflow between the project's Fluent localization files (.ftl) and GlotPress, allowing for:
- **Source generation**: Converts the source English Fluent file to PO format (.pot) for uploading to GlotPress
- **Translation sync**: Downloads translated PO files from GlotPress and converts them back to Fluent format, adding them to the project
- **Locale management**: Supports MAG16 locales with mapping between GlotPress and project locale codes; this was implemented just to limit the languages we know we support, but this mapping can be easily updated in the future

## New lanes
- `generate_source_po_file` lane - Converts `en-US/main.ftl` to `en-US.pot` for GlotPress upload
- `download_translations` lane - Downloads all available translations from GlotPress following the supported locales list and updates local Fluent files
- `generate_fluent_file_from_po` lane - Converts individual PO files back to Fluent format
- `download_po_files_from_glotpress` lane - Downloads PO files for all supported locales

## Proposed localization workflow
A day-to-day workflow can look like this:

1. New features are implemented, new localization strings need to be created (or updated).
2. Developers add strings to the source language file `wp_localization/localization/en-US/main.ftl`.
3. With a certain frequency or based on a trigger (for example, new commits to `trunk` or a daily scheduled build), we run `generate_source_po_file` and commit the updated `wp_localization/glotpress/en-US.pot`.
4. A `wpcom` job uploads the updated `en-US` source file to GlotPress (see 184627-ghe-Automattic/wpcom )
5. Again with a given frequency, we fetch the translations, generate the Fluent files and commit them to the project (or create a PR for review) by running `download_translations`.

**Update**: The PR https://github.com/Automattic/wordpress-rs/pull/770 integrates the automation using the Nightly schedule as the trigger mentioned above.

iangmaia committed:
Use latest release-toolkit Gem

iangmaia committed:
Add Fluent / PO localization lanes

iangmaia committed:
Update source PO file (en-US.pot) to be synced to GlotPress

iangmaia committed:
Update generated Fluent files based on latest GlotPress translations

iangmaia committed:
Update release-toolkit

iangmaia committed:
Fix Rubocop violations

iangmaia committed:
Run bundle update

iangmaia committed:
Create temporary dir for downloading PO files and use open-uri for downloading

iangmaia committed:
Update code to use `fluent-tools` Gem

iangmaia committed:
Update generated source PO file

iangmaia committed:
Update generated Fluent files based on latest GlotPress translations

AliSoftware commented on Gemfile:
```diff
@@ -2,5 +2,6 @@
-gem 'fastlane', '~> 2.228'
+gem 'fastlane', '~> 2.227'
 gem 'fastlane-plugin-wpmreleasetoolkit', '~> 13.2'
+gem 'fluent-tools', '~> 0.1', git: 'https://github.com/Automattic/fluent-rust-tools.git', branch: 'iangmaia/ruby-gem-support'
```
🎗️ As you mentioned in the PR description, reminder to update this once the gem is published before merging this PR

crazytonyli commented on fastlane/Fastfile:
```diff
@@ -11,6 +14,35 @@ LANE_VALUE_XCFRAMEWORK_CHECKSUM_PATH = 'WP_XCFRAMEWORK_CHECKSUM_PATH'
+  { glotpress: 'sv',    project: 'sv'     },
+  { glotpress: 'tr',    project: 'tr-TR'  },
+  { glotpress: 'zh-cn', project: 'zh-CN'  },
+  { glotpress: 'zh-tw', project: 'zh-TW'  }
```
This is shorter than I thought it'd be. The translations in this library should be a combination of the locales supported in the iOS and Android apps. Here is the list that I gathered a while ago:

```
ar
az
bg
cs
cy
da
de
el
en-AU
en-CA
en-GB
en-US
es
es-CL
es-CO
es-MX
es-VE
eu
fr
fr-CA
gd
gl
he
hi
hr
hu
id
is
it
ja
kmr
ko
lv
mk
ms
nb
nl
pl
pt
pt-BR
ro
ru
sk
sq
sr
sv
th
tr-TR
uz
vi
zh-CN
zh-HK
zh-TW
```

iangmaia replied:
I've updated the list of supported locales on 51af12c644bc9661329804742a11d0b3d9b73cac (for the record, I also had to add `fr-CA` in GlotPress). Thanks @crazytonyli!

crazytonyli commented:
Thanks for working on this! The result looks great!

AliSoftware commented on wp_localization/glotpress/en-US.pot:
💄  I wonder if it would be worth adding a `README.md` in that `wp_localization/glotpress/` folder, even a very simple one, to explain what it's for?

```markdown
### What is this folder?

We use GlotPress as a platform/service to have translators translate our strings from English to other languages.
But GlotPress does not support Rust's Fluent `.ftl` format as an input/output file format.

To circumvent that, during the localization automation process, we convert the `localization/en-US/main.ftl` file to the PO format in `glopress/en-US.pot` so that a cron job can then later pick up that `.pot` file[^1] and upload it to GlotPress and send it to translators. This transformation is handled by `bundle exec fastlane generate_source_po_file`.

Later we then download the translations from GlotPress (which are exported in the `.po` format), then regenerate the `localization/*/main.ftl` files for each language based on those downloaded translated `.po` files. This is handled by `bundle exec fastlane download_translations`

[^1]: Note that we have to commit the `glotpress/en-US.pot` file here because that file is picked up by a cron job in our systems on a regular basis (as opposed to the `.pot` being sent via API on demand), given how those imports are integrated in the company's larger localization system (with import logs and packages and pings sent to translators etc. that a simple API upload to GlotPress wouldn't cover otherwise)
```

iangmaia replied:
Good point, thanks!
I've added it to the [sync job PR](https://github.com/Automattic/wordpress-rs/pull/770) as it refers to the automated stuff, but I tried to add the process and lanes description to the README as well: d69ebf30860b43101d3a5f9be560323d048dcb67

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -105,6 +137,152 @@ lane :publish_to_s3 do
+  UI.message("✅ #{File.basename(file_path)} → #{fluent_file_path}")
+
+  fluent_file_path
+end
```
💄 / 💡  Opinionated: I'd find it easier to read if that whole helper lane was moved to after the `download_translations` and `download_po_files_from_glotpress` lanes in the `Fastfile`.

This simple move would allow the reader of the `Fastfile` to better understand that `generate_source_po_file` and `download_translations` lanes, that would then appear first, are the main lanes one is supposed to call, while `download_po_files_from_glotpress` and `generate_fluent_file_from_po` are helper lanes called by the one above it.

This also helps organizing the lanes in the `Fastfile` by the order in which they'd typically be called by the automation, with `download_translations` calling `download_po_files_from_glotpress` first then `generate_fluent_file_from_po` next on (each locale).

iangmaia replied:
Agreed! Updated in 1d9885491533c195ee1ad844bbd4b4610a0ab916

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -105,6 +137,152 @@ lane :publish_to_s3 do
+  fluent_file_path = File.join(LOCALIZATION_FLUENT_FILES_DIR, locale, MAIN_FLUENT_FILE_NAME)
+
+  UI.user_error!("❌ PO file not found: #{file_path}") unless File.exist?(file_path)
+
```
💄 Might make more sense to do the guard statement earlier, since it's checking the input param?

```suggestion
  UI.user_error!("❌ PO file not found: #{file_path}") unless File.exist?(file_path)

  locale = File.basename(file_path, '.po')
  fluent_file_path = File.join(LOCALIZATION_FLUENT_FILES_DIR, locale, MAIN_FLUENT_FILE_NAME)

```

iangmaia replied:
👍 Updated on 98947d604f76342fc5e569392685ec9fc8a074ab

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -105,6 +137,152 @@ lane :publish_to_s3 do
+    updated_fluent_files = []
+
+    # Convert PO files back to Fluent format
+    if downloaded_files.any?
```
💄  I'd suggest to instead use a guard statement to reduce indentation and exit early:

```ruby
if downloaded_files.empty?
  UI.message("No .po files were downloaded from GlotPress")
  next
end

UI.header('🔄 Converting PO files to Fluent format')

# … and the rest of the code, without the need for the extra indentation
```

iangmaia replied:
👍 Updated in 55c1118a838c16cd4dd28c1e1aedf5b852203ff9

AliSoftware commented on fastlane/Fastfile:
```diff
@@ -105,6 +137,152 @@ lane :publish_to_s3 do
+lane :download_translations do |commit_and_push_changes: false|
+  UI.header('🌐 Downloading translations from GlotPress')
+
+  Dir.mktmpdir do |temp_download_dir|
+    downloaded_files = download_po_files_from_glotpress(download_dir: temp_download_dir)
```
❤️  Nice touch using a temp dir for the download and processing here, will not only prevent any intermediate file from being in the repo (even if we could have gitignored them) and also prevent any previous runs and translations downloads from impacting future ones (e.g. a `fr.po` file downloaded on first run but not anymore on the 2nd run for any reason, which would lead to the previous `fr.po` file to not be replaced and its old content to be used if you were not using a `mktmpdir`) 👍

AliSoftware reviewed (approved):
LGTM 👍 

Left a couple of nitpicks but just mostly form and code adjustments, nothing major.
Approving to unblock (though reminder not to merge this until the `fluent-tools` gem is finalized and published as a tagged version so we can point to it instead of the branch)

iangmaia committed:
Reorder helper lane

iangmaia committed:
Validate file_path at the beginning of the `generate_fluent_file_from_po` lane

iangmaia committed:
Update code to exit early in `download_translations` when there are no downloaded files

iangmaia committed:
Add all locales supported by the iOS and Android apps
h/t to @crazytonyli for highlighting that the supported locales should be a combination of the locales supported in the iOS and Android apps https://github.com/Automattic/wordpress-rs/pull/765#discussion_r2151887208

iangmaia committed:
Add generated Fluent files with the updated list of supported locales

iangmaia committed:
Update fluent-tools to 0.2.0 and re-generate localization files with the latest version

iangmaia committed:
Update to 0.3.0 release

iangmaia committed:
Update PO file with latest fluent-tools

iangmaia committed:
Merge branch 'trunk' into iangmaia/fluent-to-glotpress-with-rust-tools
